pennylane/compiler/python_compiler/transforms/quantum/parity_synth.py

```diff
@@ -0,0 +1,353 @@
+            if i != j  #                                                  │
+        ]  #                                                              │
+    )  #                                                                  ╯
+    arbor = nx.minimum_spanning_arborescence(parity_graph)  # Line 6
+
+    # Find the root of the tree
+    root = next(iter(node for node, degree in arbor.in_degree() if degree == 0))
+
+    P = np.concatenate([P[:, :parity_idx], P[:, parity_idx + 1 :]], axis=1)  # Line 4
+    # Lines 7-10, update P and inv_synth_matrix in place
+    sub_circuit = _apply_dfs_po_circuit(arbor, root, P, inv_synth_matrix)
+    circuit.append((parity_idx, root, sub_circuit))  # Record parity index, qubit index, CNOTs
+    return P, inv_synth_matrix, circuit
+
+
+def _parity_network_synth(P: np.ndarray) -> list[int, list[tuple[int]]]:
+    """Main subroutine for the ``ParitySynth`` pass, mostly a ``for``-loop wrapper around
+    ``_loop_body_parity_network_synth``. It synthesizes the parity network, as described
+    in Algorithm 1 in https://arxiv.org/abs/2104.00934.
+
+    Args:
+        P (np.ndarray): Parity table to be synthesized.
+            Shape should be ``(num_wires, num_parities)``
+
+    Returns:
+        tuple[list[int, list[tuple[int]]], np.ndarray]: Synthesized parity network, as a
+        circuit with structure as described in ``_loop_body_parity_network_synth``. Also,
+        inverse of the parity matrix implemented by the synthesized circuit.
+
+    """
+    if P.shape[-1] == 0:
+        # Nothing to do if there are not parities
+        return [], None
+
+    circuit = []  # Line 1 in Alg. 1
+    num_wires, num_parities = P.shape
+    # Initialize an inverse parity matrix that is updated with the CNOTs that are synthesized here.
+    inv_synth_mat = np.eye(num_wires, dtype=int)
+    # `num_parities` loop iterations because each loop body takes care of one parity, we just
+    # don't know which one. This makes the `for`-loop equivalent to line 2 in Alg. 1
+    for _ in range(num_parities):
+        P, inv_synth_mat, circuit = _loop_body_parity_network_synth(P, inv_synth_mat, circuit)
+
+    return circuit, inv_synth_mat % 2
+
+
+### end of xDSL-agnostic part
+
+valid_phase_polynomial_ops = {"CNOT", "RZ"}
+
+
+def make_phase_polynomial(
+    ops: list[CustomOp],
+    init_wire_map: dict[QubitType, int],
+) -> tuple[np.ndarray]:
+    r"""Compute the phase polynomial representation of a list of ``CustomOp``\ s.
+    This implementation is very similar to :func:`~.transforms.intermediate_reps.phase_polynomial`
+    but adjusted to work with xDSL objects."""
+    wire_map = init_wire_map
+
+    parity_matrix = np.eye(len(wire_map), dtype=int)
+    parity_table = []
+    angles = []
+    arith_ops = []
+    for op in ops:
+        name = op.gate_name.data
+        if name == "CNOT":
+            control, target = wire_map.pop(op.in_qubits[0]), wire_map.pop(op.in_qubits[1])
```

Comment:
Why do we remove wires from the map for CNOT but not RZ?

Reply:
That was a bug, nice catch! I think I fixed it in a follow-up PR but not here in the root branch 😅